src/System.Net.Sockets/src/System/Net/Sockets/Socket.Unix.cs

```diff
@@ -17,5 +21,136 @@ private Socket GetOrCreateAcceptSocket(Socket acceptSocket, bool unused, string
             handle = null;
             return null;
         }
+
+        private void SendFileInternal(string fileName, byte[] preBuffer, byte[] postBuffer, TransmitFileOptions flags)
+        {
+            // Note, UseDefaultWorkerThread is the default and is == 0.
+            // Unfortunately there is no TransmitFileOptions.None.
+            if (flags != TransmitFileOptions.UseDefaultWorkerThread)
+            {
+                throw new PlatformNotSupportedException();
+            }
+
+            // Open the file, if any
+            // Open it before we send the preBuffer so that any exception happens first
+            FileStream fileStream = null;
+            if (fileName != null && fileName.Length > 0)
+            {
+                fileStream = new FileStream(fileName, FileMode.Open, FileAccess.Read, FileShare.Read);
+            }
+
+            SocketError errorCode = SocketError.Success;
+            using (fileStream)
+            {
+                // Send the preBuffer, if any
+                // This will throw on error
+                if (preBuffer != null && preBuffer.Length > 0)
+                {
+                    Send(preBuffer);
+                }
+
+                // Send the file, if any
+                if (fileStream != null)
+                {
+                    // This can throw ObjectDisposedException.
+                    errorCode = SocketPal.SendFile(_handle, fileStream);
+                }
+            }
+
+            if (errorCode != SocketError.Success)
+            {
+                SocketException socketException = new SocketException((int)errorCode);
+                UpdateStatusAfterSocketError(socketException);
+                if (NetEventSource.IsEnabled) NetEventSource.Error(this, socketException);
+                throw socketException;
+            }
+
+            // Send the postBuffer, if any
+            // This will throw on error
+            if (postBuffer != null && postBuffer.Length > 0)
+            {
+                Send(postBuffer);
+            }
+        }
+
+        private async Task SendFileInternalAsync(FileStream fileStream, byte[] preBuffer, byte[] postBuffer)
+        {
+            SocketError errorCode = SocketError.Success;
+            using (fileStream)
+            {
+                // Send the preBuffer, if any
+                // This will throw on error
+                if (preBuffer != null && preBuffer.Length > 0)
+                {
+                    // Using "this." seems to make the extension method kick in
+                    await this.SendAsync(new ArraySegment<byte>(preBuffer), SocketFlags.None);
```

Comment:
Surprisingly, the "this" does seem to be necessary.  If I remove it, I get:

`  System\Net\Sockets\Socket.Unix.cs(127,23): error CS1501: No overload for method 'SendAsync' takes 2 arguments 
`

My guess is that this is because there are non-extension SendAsync methods, and for some reason the compiler is only looking for these in the non-this case.

I'm not sure if this is a compiler bug or by design...